cloud-exec.md

```diff
@@ -0,0 +1,153 @@
+# Cloud execution
+
+- Enhancement Proposal PR: (leave this empty)
+- Contributors: dmpetrov
+
+# Summary
+
+Users need to run ML experiments in cloud instances or remote machines.
+DVC should support this out of the box to improve user experience and remove
+  the overhead of instrumenting these scenarios.
+
+# Motivation
+
+Today users need to spend time on instrumenting remote execution when there is
+  a lack of local resources for ML trainig (like GPU or RAM) or when it is a
+  company policy to train on dedicated hardware or on Kubernetes cluster.
+The existing way of executing training through [CML](cml.dev) (CI/CD) works
+  great for handing off ML models between team members but is not the best
+  approach in an experimentation phase since CI/CD creates an overhead on the
+  execution time and it creates some footprint/records in CI/CD that is
+  a bit overwhelming when the number of experiments is huge (10s, 100s or
+  1000s).
+This feature can save time for users and improve ML experimentation experience.
+
+## Scenarios
+
+1. **Run exp.** Allocate a cloud istance (with a specified configuration) and
+  execute an ML experiment `dvc exp run` (or `dvc repro`) on it.
+2. **Run stage.** Allocate a cloud istance and execute a particular stage of
+   an experiment on it.
+3. **Hot instance.** Execute on an existing running instance without instance
+   allocation.
+4. **Remote machine.** Execute in a remote machine (cloud or on-premise one).
+5. **Instance pool.** Execute on one of the existing running instances.
+6. **Exp queue** Execute a set of experiments (from exp queue for example) in
+   an instance pool or a single instance.
+7. **Web run.** Execute an experiment from SaaS\Web using one of the methods
+   from above and DVC machinery.
```

Comment:
Are there different categories of scenarios? Some of these overlap. For example, a web run could apply to any of the other scenarios. To support all of these, it seems like these specs would be needed:

### Compute resource spec
**Provisioning:** New vs. existing
**Location:** Cloud vs. on-premises
**Count:** Single instance vs pool

### DVC execution spec
**Stages**: Single stage vs. full pipeline
**Experiments**: Single vs. multiple
**Entrypoint**: Local vs. SaaS/web

Does that seem accuratre? And is that a useful way to organize it?

Reply:
@dberenbaum I think the scenarios can def. be grouped. This is probably a non-exhaustive list of use cases.

I like the categories you summarized but wouldn't see them as specs (if you meant specs to implement the feature). The user will prob care about these distinctions, but they may not be relevant for the tool except perhaps whether running on a single or several instances (IF this includes parallel execution).

This is great for how to teach this, though!

Reply:
@dberenbaum some regrouping might be needed. But in some cases, such grouping might be not relevant as @jorgeorpinel mentioned. An example - on-premise run is a prerequisite of cloud. I'd keep the whole list for now and try to cut the scope when we start implementing this.

Reply:
No problem, this just helps me organize it better in my head.